### ISSUE TITLE
Changes to drop volatile lead to kokkos-kernels nightly test failures with OpenMP, Cuda backends on Volta70

### ISSUE THREAD

**Issue by ndellingwood** [kokkos/kokkos#4941]
The changes to drop volatile from the reduce and scan join routines with PR #4931 results in failure of some kokkos-kernels tests (graph_openmp, graph_cuda, sparse_cuda) with OpenMP and Cuda backends on Power9+Volta70 arch:

graph_openmp: fails with gcc/6.4.0, gcc/7.2.0
graph_cuda, sparse_cuda: fails with cuda/9.2.88+gcc/7.2.0

Tests fail with SHA 335468d25c279ebbfea2f9914800c2830d08714c
Tests pass with SHA 9429e13 (merge commit prior to #4931)

kokkos-kernels SHA kokkos/kokkos-kernels@403a4177774fda29426e289367de705bf9cbe0a8

More detailed summary,
graph_openmp:
```

5: [----------] Global test environment tear-down
5: [==========] 18 tests from 1 test case ran. (23254 ms total)
5: [  PASSED  ] 8 tests.
5: [  FAILED  ] 10 tests, listed below:
5: [  FAILED  ] openmp.graph_graph_color_deterministic_double_int_int_TestExecSpace
5: [  FAILED  ] openmp.graph_graph_color_deterministic_double_int_size_t_TestExecSpace
5: [  FAILED  ] openmp.graph_graph_color_distance2_double_int_int_TestExecSpace
5: [  FAILED  ] openmp.graph_graph_color_bipartite_sym_double_int_int_TestExecSpace
5: [  FAILED  ] openmp.graph_graph_color_bipartite_row_double_int_int_TestExecSpace
5: [  FAILED  ] openmp.graph_graph_color_bipartite_col_double_int_int_TestExecSpace
5: [  FAILED  ] openmp.graph_graph_color_distance2_double_int_size_t_TestExecSpace
5: [  FAILED  ] openmp.graph_graph_color_bipartite_sym_double_int_size_t_TestExecSpace
5: [  FAILED  ] openmp.graph_graph_color_bipartite_row_double_int_size_t_TestExecSpace
5: [  FAILED  ] openmp.graph_graph_color_bipartite_col_double_int_size_t_TestExecSpace
```

**Reproducer** (weaver, gcc/7.2.0):
```
# rhel7W queue
module load gcc/7.2.0 cmake/3.19.3

$KOKKOSKERNELS_PATH/cm_generate_makefile.bash --with-devices=OpenMP --arch=Power9,Volta70 --compiler=g++ --cxxflags="-O3 -Wall -Wunused-parameter -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized " --cxxstandard="14" --kokkos-path=$KOKKOS_PATH --kokkoskernels-path=$KOKKOSKERNELS_PATH --with-scalars='double,complex_double' --with-ordinals=int --with-offsets=int,size_t --with-layouts=LayoutLeft --no-examples
```

@PhilMiller are you able to investigate this impact of #4931 on kokkos-kernels ?
Adding @lucbv @e10harvey so they are in the loop, this will impact testing of incoming PRs

**Comment by ndellingwood** [comment]
@PhilMiller in case useful extra info, the `graph_openmp` failures in kokkos-kernels also occur with OpenMP builds on ARMV8_THUNDERX2, tested with armclang/20.1 on the Mayer testbed

**Comment by PhilMiller** [comment]
> @PhilMiller in case useful extra info, the `graph_openmp` failures in kokkos-kernels also occur with OpenMP builds on ARMV8_THUNDERX2, tested with armclang/20.1 on the Mayer testbed

Cool, I'll be able to easily test that on Stria

**Comment by ndellingwood** [comment]
In the Cuda builds and `sparse_cuda` tests, here is a snip of output from failing spgemm test:

```
20:56:57 4: [ RUN      ] cuda.sparse_spgemm_double_int_int_TestExecSpace
20:56:57 4: /data/jenkins-new/workspace/KokkosKernels_semsrhel7gpu01_cuda111_gcc830/kokkos-kernels/unit_test/sparse/Test_Sparse_spgemm.hpp:429: Failure
20:56:57 4: Value of: success
20:56:57 4:   Actual: false
20:56:57 4: Expected: true
20:56:57 4: KKMEM still has issue 402 bug! Error message:
20:56:57 4: Kokkos failed to allocate memory for label "pool data".  Allocation using MemorySpace named "Cuda" failed with the following error:  Allocation of size 1.718e+10 G failed, likely due to insufficient memory.  (The allocation mechanism was cudaMalloc().  The Cuda allocation returned the error code ""cudaErrorMemoryAllocation".)
20:56:57 4: 
20:56:57 4: 
20:56:57 4: entries are different.
20:56:57 4: 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 ... ... ... 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
20:56:57 4: 0 2 45 59 130 171 259 273 321 346 347 353 463 464 495 517 669 711 727 988 ... ... ... 1793 1794 1795 1796 1797 1798 1799 1800 1801 1802 1803 1804 1805 1806 1807 1808 1809 1810 1811 1812 
20:56:57 4: /data/jenkins-new/workspace/KokkosKernels_semsrhel7gpu01_cuda111_gcc830/kokkos-kernels/unit_test/sparse/Test_Sparse_spgemm.hpp:432: Failure
20:56:57 4: Value of: correctResult
20:56:57 4:   Actual: false
20:56:57 4: Expected: true
20:56:57 4: KKMEM still has issue 402 bug; C=AA' is incorrect!
20:56:57 4: 
20:56:57 4: [  FAILED  ] cuda.sparse_spgemm_double_int_int_TestExecSpace (6293 ms)
```

That was a build with cuda/11.1 on x86+Volta70, I'm assuming this could be reproduced on kokkos-dev-2

**Reproducer** (Cuda+Serial)
```
module load sems-archive-env sems-archive-cmake/3.17.1 sems-archive-gcc/8.3.0 sems-archive-cuda/11.1

$KOKKOSKERNELS_PATH/cm_generate_makefile.bash --with-devices=Cuda,Serial --arch=Volta70 --compiler=$KOKKOS_PATH/bin/nvcc_wrapper --cxxflags="-O3 -Wall -Wunused-parameter -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized " --cxxstandard="14" --kokkos-path=$KOKKOS_PATH --kokkoskernels-path=$KOKKOSKERNELS_PATH --with-scalars='double,complex_double' --with-ordinals=int --with-offsets=int,size_t --with-layouts=LayoutLeft --with-cuda-options=enable_lambda   --no-examples
```

**Comment by ndellingwood** [comment]
The Cuda build I referenced above https://github.com/kokkos/kokkos/issues/4941#issuecomment-1091049307 was running at the same time as another job (the system tested on does not have dedicated compute nodes), I'll shuffle the start time of the test and update after another run in case the failure is due to jobs trying to execute tests simultaneously

**Comment by PhilMiller** [comment]
The enormous size of the reported failed allocation on CUDA is very suspicious. That could arise from a bad reduction result, I guess, but the values should come out too low, not outrageously high

**Comment by ndellingwood** [comment]
I relaunched the cuda/11 build to ensure it was the sole job running on the machine, the failures with spgemm https://github.com/kokkos/kokkos/issues/4941#issuecomment-1091049307 remained and failed in a debug build as well

**Comment by PhilMiller** [comment]
https://github.com/kokkos/kokkos-kernels/pull/1382 @ndellingwood That PR isn't the fix, but it will be helpful to making the fix stick.

**Comment by srajama1** [comment]
@PhilMiller @ndellingwood Can we revert the commit that caused this issue, fix everything and push the changes in? This is blocking milestone work in Kokkos Kernels side.

**Comment by ndellingwood** [comment]
@PhilMiller after the changes in #4931 merged (only change from previous night's passing build), this `-Werror` also showed up in some kokkos-kernels builds (gcc/8.3.0 here):

```
01:13:16 In file included from /home/jenkins/jenkins-new/workspace/KokkosKernels_KokkosDev2_GCC830/kokkos-kernels/src/batched/dense/impl/KokkosBatched_QR_WithColumnPivoting_TeamVector_Internal.hpp:8,
01:13:16                  from /home/jenkins/jenkins-new/workspace/KokkosKernels_KokkosDev2_GCC830/kokkos-kernels/src/batched/dense/impl/KokkosBatched_QR_WithColumnPivoting_TeamVector_Impl.hpp:7,
01:13:16                  from /home/jenkins/jenkins-new/workspace/KokkosKernels_KokkosDev2_GCC830/kokkos-kernels/src/batched/dense/KokkosBatched_QR_WithColumnPivoting_Decl.hpp:28,
01:13:16                  from /home/jenkins/jenkins-new/workspace/KokkosKernels_KokkosDev2_GCC830/kokkos-kernels/unit_test/batched/dense/Test_Batched_TeamVectorQR_WithColumnPivoting.hpp:12,
01:13:16                  from /home/jenkins/jenkins-new/workspace/KokkosKernels_KokkosDev2_GCC830/kokkos-kernels/unit_test/batched/dense/Test_Batched_Dense.hpp:85,
01:13:16                  from /home/jenkins/jenkins-new/workspace/KokkosKernels_KokkosDev2_GCC830/kokkos-kernels/unit_test/threads/Test_Threads_Batched_Dense.cpp:5:
01:13:16 /home/jenkins/jenkins-new/workspace/KokkosKernels_KokkosDev2_GCC830/kokkos-kernels/src/batched/dense/impl/KokkosBatched_FindAmax_Internal.hpp: In static member function ‘static int KokkosBatched::TeamVectorQR_WithColumnPivotingInternal::invoke(const MemberType&, int, int, ValueType*, int, int, ValueType*, int, IntType*, int, ValueType*, int&) [with MemberType = Kokkos::Impl::ThreadsExecTeamMember; ValueType = double; IntType = int]’:
01:13:16 /home/jenkins/jenkins-new/workspace/KokkosKernels_KokkosDev2_GCC830/kokkos-kernels/src/batched/dense/impl/KokkosBatched_FindAmax_Internal.hpp:58:55: error: ‘*((void*)& value +8)’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
01:13:16        Kokkos::single(Kokkos::PerTeam(member), [&]() { *idx = value.loc; });
01:13:16                                                        ^
01:13:16 /home/jenkins/jenkins-new/workspace/KokkosKernels_KokkosDev2_GCC830/kokkos-kernels/src/batched/dense/impl/KokkosBatched_FindAmax_Internal.hpp:46:26: note: ‘*((void*)& value +8)’ was declared here
01:13:16        reducer_value_type value;
```

**Comment by PhilMiller** [comment]
@ndellingwood With kokkos as in #4951 and kokkos-kernels `develop` I couldn't reproduce that failure on gcc 8.3.0 with builds following either of the following command lines:
```
~/repos/kokkos-kernels/cm_generate_makefile.bash --with-devices=OpenMP,Serial --cxxflags="-O3 -Wall -Wunused-parameter -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized -Wmaybe-uninitialized " --cxxstandard="14" --kokkos-path=$HOME/repos/kokkos --kokkoskernels-path=$HOME/repos/kokkos-kernels --with-scalars='double,complex_double' --with-ordinals=int --with-offsets=int,size_t --with-layouts=LayoutLeft
```

```
~/repos/kokkos-kernels/cm_generate_makefile.bash --with-devices=OpenMP,Serial --cxxflags="-g -Wall -Wunused-parameter -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized -Wmaybe-uninitialized " --cxxstandard="14" --kokkos-path=$HOME/repos/kokkos --kokkoskernels-path=$HOME/repos/kokkos-kernels --with-scalars='double,complex_double' --with-ordinals=int --with-offsets=int,size_t --with-layouts=LayoutLeft
```

Both also passed their tests.

I'm waiting on some other stuff to build, and then I'll test against kokkos `develop` to confirm that I can reproduce it at all, and that the above PR ends up fixing (quieting) it

**Comment by PhilMiller** [comment]
I can't reproduce that with kokkos `develop` either, with either configuration. I'll need some more detail to run that one down.

**Comment by PhilMiller** [comment]
I was able to reproduce the reported test failures on OpenMP (`graph_openmp`) and Cuda (`graph_cuda`, `sparse_cuda`), and confirmed that they pass with #4951.

**Comment by PhilMiller** [comment]
Fix merged, bug was never in a release, so closing.

**Comment by ndellingwood** [comment]
Thanks @PhilMiller !

**Comment by ndellingwood** [comment]
@PhilMiller @dalg24 with #4951 merged I'm seeing compilation errors (-Werror=maybe-uninitialized) in kokkos-kernels with gcc/7.2.0,

```
In file included from /home/jenkins/blake-new/workspace/KokkosKernels_PullRequest_GCC720/kokkos-kernels/src/batched/dense/impl/KokkosBatched_QR_WithColumnPivoting_TeamVector_Internal.hpp:8:0,
                 from /home/jenkins/blake-new/workspace/KokkosKernels_PullRequest_GCC720/kokkos-kernels/src/batched/dense/impl/KokkosBatched_QR_WithColumnPivoting_TeamVector_Impl.hpp:7,
                 from /home/jenkins/blake-new/workspace/KokkosKernels_PullRequest_GCC720/kokkos-kernels/src/batched/dense/KokkosBatched_QR_WithColumnPivoting_Decl.hpp:28,
                 from /home/jenkins/blake-new/workspace/KokkosKernels_PullRequest_GCC720/kokkos-kernels/unit_test/batched/dense/Test_Batched_TeamVectorQR_WithColumnPivoting.hpp:12,
                 from /home/jenkins/blake-new/workspace/KokkosKernels_PullRequest_GCC720/kokkos-kernels/unit_test/batched/dense/Test_Batched_Dense.hpp:85,
                 from /home/jenkins/blake-new/workspace/KokkosKernels_PullRequest_GCC720/kokkos-kernels/unit_test/threads/Test_Threads_Batched_Dense.cpp:5:
/home/jenkins/blake-new/workspace/KokkosKernels_PullRequest_GCC720/kokkos-kernels/src/batched/dense/impl/KokkosBatched_FindAmax_Internal.hpp: In static member function static int KokkosBatched::TeamVectorQR_WithColumnPivotingInternal::invoke(const MemberType&, int, int, ValueType*, int, int, ValueType*, int, IntType*, int, ValueType*, int&) [with MemberType = Kokkos::Impl::ThreadsExecTeamMember; ValueType = double; IntType = int]:
/home/jenkins/blake-new/workspace/KokkosKernels_PullRequest_GCC720/kokkos-kernels/src/batched/dense/impl/KokkosBatched_FindAmax_Internal.hpp:58:55: error: value.Kokkos::ValLocScalar::loc may be used uninitialized in this function [-Werror=maybe-uninitialized]
       Kokkos::single(Kokkos::PerTeam(member), [&]() { *idx = value.loc; });
                                                       ^
/home/jenkins/blake-new/workspace/KokkosKernels_PullRequest_GCC720/kokkos-kernels/src/batched/dense/impl/KokkosBatched_FindAmax_Internal.hpp:46:26: note: value.Kokkos::ValLocScalar::loc was declared here
       reducer_value_type value;
                          ^~~~~
```

Should I open a new issue or reopen this issue?

**Comment by PhilMiller** [comment]
Please open a fresh issue, with complete reproducer instructions. As noted above, I tried adding `-Wmaybe-uninitialized` to my builds, and didn't see that come up.
